Could GOTURN be run with CPU only?

### I find simply uncomment some lines related to Caff and CUDA in CMakeLists.txt,then I met some fatal error like "/home/wxb/git/caffe/GOTURN/src/network/regressor.h:4:27: fatal error: caffe/caffe.hpp: No such file or directory". I guess GOTURN can not be run without GPU and CUDA installed.

But I found some information in your paper:
"On an Nvidia GeForce GTX Titan X GPU with cuDNN acceleration, our tracker runs at 6.05 ms per frame (not including the 1 ms to load each image in OpenCV), or 165 fps. On a GTX 680 GPU, our tracker runs at an average of 9.98 ms per frame, or 100 fps. **If only a CPU is available, the tracker runs at 2.7 fps.**"

Look at the bold text, does that mean GOTURN can be run with CPU only?If that's true, what should I do?

Thank you!

@davheld Hi, finally I solved all errors and now I can run GOTURN with CPU only on my ubuntu 16.04LTD. Here are my advices:
   （1）To run with CPU only, uncomment CPU_ONLY=1 and TEST_GPUID := 0, and comment all items related to CUDA. Then compile Caffe with "cmake -DCPU_ONLY=ON .." in /build directory, if not , you may find CPU_ONLY equal OFF in the return information of cmake . Also you should add "#define CPU_ONLY 1" to all header files in /src/network directory. Beside , I also follow @davheld's advice and add it to other main cpp files in other directory like /train.
   （2）After Caffe is successfully installed and tested, firstly install trax (#2) (just download and make it), or you will meet errors like "undefined reference to `trax_server_setup" during make. Then configure GOTURN/CMakeLists.txt like below:

![screenshot from 2017-03-03 11-52-17](https://cloud.githubusercontent.com/assets/17294093/23537563/e0086ba2-0007-11e7-9043-625352386bab.png)
![screenshot from 2017-03-03 11-53-15](https://cloud.githubusercontent.com/assets/17294093/23537589/063802c4-0008-11e7-8aeb-9e6c950751ba.png)
(Caff_LIBRARIES should be set to /path-to-caffe/build/install/lib/libcaffe.so, not /path-to-caffe/build/install/lib. If not , errors like "Target:...requests linking to directory ...Targets may link only to libraries.  CMake is dropping")

and congiure GOTURN/cmake/Modules/FindCaffe.cmake like this:

![screenshot from 2017-03-03 11-55-58](https://cloud.githubusercontent.com/assets/17294093/23537637/61e52318-0008-11e7-93f8-6e7ff4531f5a.png)

   （3）Finally, build GOTURN using "cmake -DCPU_ONLY=ON .." and then make it.
Maybe some statments are reduntant, but I meet a hell lot of errors before I configure these two files right. @davheld gives me a lot help, thank you again!

--- a/.gitignore
+++ b/.gitignore
@@ -1,0 +1,6 @@
+# .gitignore
+
+*~
+build/
+nets/models/
+nets/tracker_output/

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,24 +8,34 @@ endif()
 
 find_package(Boost COMPONENTS system filesystem regex REQUIRED)
 
+
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
 find_package(TinyXML REQUIRED)
+
 
 find_package( OpenCV REQUIRED )
 message("Open CV version is ${OpenCV_VERSION}")
 
 find_package(CUDA REQUIRED)
-include_directories(${CUDA_INCLUDE_DIRS})
 # Note: If can't find CUDA, please uncomment the below line and set the path manually
 # set(CUDA_INCLUDE_DIRS /path_to_cuda/include)
+include_directories(${CUDA_INCLUDE_DIRS})
+message("CUDA_INCLUDE_DIRS is ${CUDA_INCLUDE_DIRS}")
+
 
 find_package(Caffe REQUIRED)
-include_directories(${Caffe_INCLUDE_DIRS})
-add_definitions(${Caffe_DEFINITIONS})    # ex. -DCPU_ONLY
-message("Caffe_DIR is ${Caffe_DIR}")	#specify Caffe_DIR in /cmake/Modules/FindCaffe.cmake
-# Note: If can't find Caffe, please uncomment the below line and set the path manually
+# If Caffe not found, configure line 5 of cmake/Modules/FindCaffe.cmake
+# If that fails uncomment the two lines below and set paths manually
 # set(Caffe_DIR /path_to_caffe/build/install)
 # set(Caffe_INCLUDE_DIRS /path_to_caffe/build/install/include)
+include_directories(${Caffe_INCLUDE_DIRS})
+# Uncomment for CPU only:
+# set(Caffe_DEFINITIONS -DCPU_ONLY)
+add_definitions(${Caffe_DEFINITIONS})
+message("Caffe_DEFINITIONS is ${Caffe_DEFINITIONS}")
+message("Caffe_DIR is ${Caffe_DIR}")
+message("Caffe_INCLUDE_DIRS is ${Caffe_INCLUDE_DIRS}")
+
 
 set(GLOG_LIB glog)
 

--- a/README.md
+++ b/README.md
@@ -64,6 +64,14 @@ sudo apt-get install libopencv-dev
 sudo apt-get install libtinyxml-dev 
 ```
 
+* CPU_ONLY mode (optional and not recommended)
+
+GPUs are strongly suggested as the code runs ~30X slower in CPU_ONLY mode. It is listed as an option only if your system doesn’t have GPU support. To configure this, uncomment the `CPU_ONLY` line in `CMakeLists.txt` as shown below:
+```
+# Uncomment for CPU only:
+set(Caffe_DEFINITIONS -DCPU_ONLY)
+```
+
 ### Compile
 
 From the main directory, type:

--- a/cmake/Modules/FindCaffe.cmake
+++ b/cmake/Modules/FindCaffe.cmake
@@ -1,9 +1,8 @@
 # Caffe package
 unset(Caffe_FOUND)
 
-###Set the variable Caffe_DIR as the root of your caffe directory
-#set(Caffe_DIR /*Specify path to caffe here*/)
-
+### Set the variable Caffe_DIR as the root of your caffe directory
+# set(Caffe_DIR /path_to_caffe/build/install)
 
 find_path(Caffe_INCLUDE_DIRS NAMES caffe/caffe.hpp caffe/common.hpp caffe/net.hpp caffe/proto/caffe.pb.h caffe/util/io.hpp caffe/vision_layers.hpp
   HINTS
@@ -15,7 +14,8 @@ find_library(Caffe_LIBRARIES NAMES caffe
   HINTS
   ${Caffe_DIR}/build/lib)
 
-message("lib_dirs:${Caffe_LIBRARIES}")
+message("Caffe_LIBRARIES HINTS is ${Caffe_DIR}/build/lib")
+message("Caffe_LIBRARIES is ${Caffe_LIBRARIES}")
 
 if(Caffe_LIBRARIES AND Caffe_INCLUDE_DIRS)
     set(Caffe_FOUND 1)

--- a/src/tracker/tracker_manager.cpp
+++ b/src/tracker/tracker_manager.cpp
@@ -38,8 +38,14 @@ void TrackerManager::TrackAll(const size_t start_video_num, const int pause_val)
     tracker_->Init(image_curr, bbox_gt, regressor_);
 
     // Iterate over the remaining frames of the video.
-    for (size_t frame_num = first_frame + 1; frame_num < video.all_frames.size(); ++frame_num) {
-
+    printf("Frames: ");
+    size_t frame_num = first_frame + 1;
+    for (; frame_num < video.all_frames.size(); ++frame_num) {
+      if (frame_num % 100 == 0) {
+          // force flush as printf without newline will buffer
+          printf("%lu, ", frame_num);
+          fflush(stdout);
+      }
       // Get image for the current frame.
       // (The ground-truth bounding box is used only for visualization).
       const bool draw_bounding_box = false;
@@ -63,6 +69,8 @@ void TrackerManager::TrackAll(const size_t start_video_num, const int pause_val)
       ProcessTrackOutput(frame_num, image_curr, has_annotation, bbox_gt,
                            bbox_estimate_uncentered, pause_val);
     }
+    printf("%lu\n", frame_num);
+
     PostProcessVideo();
   }
   PostProcessAll();

